Add parentheses around conditional expression in interpolated string code fix provider

MaStr11 opened the pull request:
### Customer scenario

Code fix for a conditional expression in interpolated string. Adds parenthesis around the expression.

Before (invalid code with compiler error CS8361):
```CSharp
$"{1 / Environment.TickCount == 0 ? 1 : 2}"
```
After:
```CSharp
$"{(1 / Environment.TickCount == 0 ? 1 : 2 )}"
```
This is a code fix for compiler error ERR_ConditionalInInterpolation = CS8361 introduced in  #23265.

### Bugs this fixes

#23323 

### Workarounds, if any

Add parenthesis by hand.

### Risk

Low. Specialized code fix for a very narrow compiler error.

### Performance impact

Low. Specialized code fix for a very narrow compiler error.

### Is this a regression from a previous update?

No.

### Root cause analysis

New feature.

### How was the bug found?

The need to add parenthesis caused confusion in the past. CS8361 was therefore introduced to improve the story. This is a follow up to further enhance the situation.

### Test documentation updated?

MaStr11 commented:
I need some help with at least two cases where the current solution is not sufficient because more than just the `InterpolationSyntax` needs to be replaced. That's the case if the syntax after the colon contains either double quotes or a closing curly bracket. See [TestAddParenthesisFailing_ClosingBracketInFalseCondition](https://github.com/dotnet/roslyn/pull/23587/files#diff-6913cbc746ddc9fbf263c68d3a27820fR85) and [TestAddParenthesisFailing_StringLiteralInFalseCondition](https://github.com/dotnet/roslyn/pull/23587/files#diff-6913cbc746ddc9fbf263c68d3a27820fR94) for details.
The problem is, that the SyntaxTree in those cases is totally bogus and it is hard to tell were the closing parenthesis should go. I suspect that it is not possible to cover all possible cases here but it should be defined what cases at least should work. I think something like
```
var s = $"{ cond ? "1" : "2" }";
```
is just to common to not properly support it. Supporting those cases renders the current approach useless as there is not just a single SyntaxNode.

sharwell commented:
The proposed approach of inserting the opening parenthesis and retrying the parse might be slow, but I do not believe it will be a problem. CS8361 is a compiler error, so instances of this will not be building up over time.

👍 You were correct that the parse trees are _terrible_ for these cases. Good find on a workable approach.

sharwell commented:
To @pdelvo @vweijsters @SLaks: Martin (@MaStr11) has requested examples of other edge cases related to the use of unparenthesized ternary expressions string interpolation which may be challenging to automatically correct with this code fix. Are any of you able to provide examples?

MaStr11 commented:
I changed the logic in commit c249af0 as proposed and it works surprisingly well for a lot of cases (even those with CS1026 present before the fix is applied). I wanted to share the solution in case anyone wants to pull it to find edge cases. I well add tests next week.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+                context.RegisterCodeFix(new AddParenthesisCodeAction(context.Document, conditionalExpressionSyntax.SpanStart), context.Diagnostics);
+            }
+        }
+
```
extra newline.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddParenthesisAroundConditionalExpressionInInterpolatedString), Shared]
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider : CodeFixProvider
+    {
+        private const string ERR_ConditionalInInterpolation = "CS8361"; //A conditional expression cannot be used directly in a string interpolation because the ':' ends the interpolation.Parenthesize the conditional expression.
```
Convention is the codebase is to do ```private const string CS8361 = nameof(CS8361);```

CyrusNajmabadi commented:
Hi @MaStr11, thanks for the contribution!  Definitely will be a nice feature to help people out when they run into this issue.

That said, this PR contains many code patterns that deviate from the way the rest of hte Roslyn code actions are written.  I'm going to help point them out, and would appreciate it if they could be addressed.  This isn't meant to be annoying, but only to ensure that the codebase stays nice and consistent.  Thanks!

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider
+    {
+        private sealed class AddParenthesisCodeAction : CodeAction
```
General pattern is to not create a code action with logic in it.  But instead just subclass CodeAction.DocumentChangedCodeAction and just supply an appropriate lambda that does the actual work.  See other code actions for examples of this.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.ConditionalExpressionInStringInterpolation
+{
+
```
remvoe extra newline.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+{
+
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.AddParenthesisAroundConditionalExpressionInInterpolatedString), Shared]
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider : CodeFixProvider
```
Derive from SyntaxEditorBasedCodeFixProvider.  That way you can easily supply a fast and efficient 'FixAll' experience.

MaStr11 replied:
FixAll is IMHO not useful for this kind of error (this fixes a syntax error and I don't see how there could be more then one instance of this error at a time). If there are other reasons I will change the base class (or if you just insists).

CyrusNajmabadi replied:
You're def right that fix-all is less useful here.  where it could be valuable is for someone who changes a bunch of code, or copy  pastes a lot.  But if you don't want a fix-all, that's fine :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var token = root.FindToken(diagnosticSpan.Start);
+            var conditionalExpressionSyntax = token.GetAncestor<ConditionalExpressionSyntax>();
+            var interpolationSyntax = token.GetAncestor<InterpolationSyntax>();
```
could just call GetAncesotor on the conditionalExpressionSyntax  here.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+            var diagnostic = context.Diagnostics.First();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var token = root.FindToken(diagnosticSpan.Start);
+            var conditionalExpressionSyntax = token.GetAncestor<ConditionalExpressionSyntax>();
```
convention is to not use 'Syntax' in the name of the variable.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+            var interpolationSyntax = token.GetAncestor<InterpolationSyntax>();
+            if (conditionalExpressionSyntax != null && interpolationSyntax != null)
+            {
+                context.RegisterCodeFix(new AddParenthesisCodeAction(context.Document, conditionalExpressionSyntax.SpanStart), context.Diagnostics);
```
please wrap lines that are longer than github's width.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -53,5 +53,6 @@ internal static class PredefinedCodeFixProviderNames
         public const string UseObjectInitializer = nameof(UseObjectInitializer);
         public const string UseThrowExpression = nameof(UseThrowExpression);
         public const string PreferFrameworkType = nameof(PreferFrameworkType);
+        public const string AddParenthesisAroundConditionalExpressionInInterpolatedString = nameof(AddParenthesisAroundConditionalExpressionInInterpolatedString);
```
Please keep this list sorted.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            await TestInMethodAsync(
+                @"var s = $""{ true ? ""1"" [|:|] ""2"" }"";",
+                @"var s = $""{ (true ? ""1"" : ""2"" )}"";");
```
Needs fix all test.

MaStr11 replied:
I plan to add (lots) of tests next week (I rushed this out to get feedback for the nasty string problem described in the PR comments). But I don't think FixAll is useful for this CodeFixProvider.

CyrusNajmabadi replied:
Fair enough :)

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -189,6 +189,7 @@ public static class Features
             public const string XmlTagCompletion = nameof(XmlTagCompletion);
             public const string CodeActionsAddOverload = "CodeActions.AddOverloads";
             public const string CodeActionsAddNew = "CodeActions.AddNew";
+            public const string CodeActionsAddParenthesis = nameof(CodeActionsAddParenthesis);
```
Please keep this list sorted.

CyrusNajmabadi replied:
this sholud be: ```public const string CodeActionsAddParenthesis = "CodeActions.AddParenthesis";``` so that the tests show up properly in the xunit runner.

CyrusNajmabadi replied:
This should be CodeActionsAddParenthesisAroundXXX to match the name of your type.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,48 @@
+            {
+                var text = await Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var openParenthesisPosition = ConditionalExpressionSyntaxStartPosition;
+                var textWithOpenParenthesis = text.Replace(openParenthesisPosition, 0, "(");
```
it woudl be more appropriate to just fix up teh syntax tree, rather than manually making text edits.  You could just get the InterpolationSyntax (that you got back when you did hte analysis) and wrap it's contained expression with a ParenthesizedExpression.

MaStr11 replied:
That is what I had [previously](https://github.com/dotnet/roslyn/pull/23587/commits/d0bdaf769bb4ea57a26a0be8e326bd41f051eb1e#diff-688e183cf41536b4fde1bbb91af4224fR27). See commit d0bdaf769bb4ea57a26a0be8e326bd41f051eb1. This worked well in many cases but was problematic for string literals in the false condition because the resulting syntax trees are bogus and hard to transform. So @sharwell and I decided to give this unusal approach a try (See also comments above).

CyrusNajmabadi replied:
Gotcha!  Ok.  Let's go with this approach then :)

pdelvo commented:
@sharwell I think adding parenthesis rather than removing them is uncritical. I also checked the stylecop source to look for special cases we check when adding parenthesis and I couldn't find any (except for spacing).

One thing that I think should be tested that is not covered yet is what happens if you have nested conditional expressions (shame on you if you actually do that).

MaStr11 commented:
@CyrusNajmabadi I implemented your requested changes. It would be nice to get another review. The actual code fix section starting [here](https://github.com/dotnet/roslyn/pull/23587/files#diff-688e183cf41536b4fde1bbb91af4224fR31) wasn't reviewed last time because you rejected the approach altogether, so that sections would be of special interest. I'm wondering whether that could be implement more efficient in the sense of limiting the work the compiler does to a minimum to get the CS1026 diagnostic.
Furthermore I'm not sure whether it is still useful to split the CodeFix into two files and if how these should be named.
@pdelvo Thanks for the hint about nested conditional expressions. I added the according tests.

CyrusNajmabadi commented:
Reviewing.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Traits.cs:
```diff
@@ -37,7 +37,10 @@ public static class Features
+            public const string CodeActionsAddNew = "CodeActions.AddNew";
+            public const string CodeActionsAddOverload = "CodeActions.AddOverloads";
             public const string CodeActionsAddParameter = "CodeActions.AddParameter";
+            public const string CodeActionsAddParenthesisAroundConditionalExpressionInInterpolatedString = "CodeActions.CodeActionsAddParenthesisAroundConditionalExpressionInInterpolatedString";
```
Redundant: "CodeActions.CodeActions..."

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.resx:
```diff
@@ -524,4 +524,7 @@
   <data name="Use_is_null_check" xml:space="preserve">
     <value>Use 'is null' check</value>
   </data>
+  <data name="AddParenthesisAroundConditionalExpressionInInterpolatedString" xml:space="preserve">
```
name is formed by taking the english text and replacing non-alphanumerics with _.  So in this case it would be "Add_parenthesis".

Also, this should be plural.  So "Add parentheses".

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.ConditionalExpressionInStringInterpolation
+{
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider
```
Why is this a partial type?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,37 @@
+            var conditionalExpression = token.GetAncestor<ConditionalExpressionSyntax>();
+            if (conditionalExpression != null)
+            {
+                var documentChangeAction = new CodeAction.DocumentChangeAction(
```
You'll need to make your own subclass of this type in this file (it's necessary for how our telemetry system works).  But that type will be almost empty.  You can look at other refactorings/fixes to see our general pattern here.  Just look for MyCodeAction for examples of this.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/PredefinedCodeFixProviderNames.cs:
```diff
@@ -8,6 +8,7 @@ internal static class PredefinedCodeFixProviderNames
         public const string AddAwait = nameof(AddAwait);
         public const string AddAsync = nameof(AddAsync);
         public const string AddParameter = nameof(AddParameter);
+        public const string AddParenthesisAroundConditionalExpressionInInterpolatedString = nameof(AddParenthesisAroundConditionalExpressionInInterpolatedString);
```
All these should be updated to AddParenthesesXXX (plural).

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,46 @@
+{
+    internal partial class CSharpAddParenthesisAroundConditionalExpressionInInterpolatedStringCodeFixProvider
+    {
+        private const string CS1026 = nameof(CS1026); // ) expected
```
oh.. i see this is for something else.  n/m.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,46 @@
+            // 1. Inserts an opening parenthesis
+            // 2. Re-parses the resulting document (now the colon isn't treated as starting a FormatClause anymore)
+            // 3. Looks for CS1026: ) expected
+            // 4. Inserts a closing parenthesis at CS1026
```
wow!  that's gnarly :)

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,46 @@
+            var documentWithOpenParenthesis = document.WithText(textWithOpenParenthesis);
+            var syntaxTree = await documentWithOpenParenthesis.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxRoot = await syntaxTree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+            var conditionalExpression = syntaxRoot.FindNode(new TextSpan(openParenthesisPosition, 0));
```
this won't be a conditional expression right?  because you added the parenthesis, wont' this be a parenthesized expression?

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/AddParenthesisCodeAction.cs:
```diff
@@ -0,0 +1,46 @@
+                return documentWithOpenParenthesis.WithText(textWithBothParenthesis);
+            }
+
+            return documentWithOpenParenthesis;
```
Oh... this seems weird.  We could get in a state where we only added the open paren?

MaStr11 replied:
Yes. See this [test](https://github.com/dotnet/roslyn/pull/23587/files#diff-6913cbc746ddc9fbf263c68d3a27820fR293).

CyrusNajmabadi replied:
That seems like a super bizarre case to handle.  The user didn't provide parens around the conditional, but they did supply a close paren?  I'd be happy to not care about that case at all :)

The real scenario is that people will type ```$"foo {bar ? baz : quux} ztech"``` and then will get the compiler error.  We need to just fix that up, and we'll be good :)

CyrusNajmabadi commented:
Ok.  Sorry to ask this, but i think this needs an slight change to the implementation.  Specifically, i don't like that we can end up trying to fix things, but then might run into a problem half way through the fix that then means we do a partial fix.

To address this, you can do the following:

Move the actual attempt to fix the document into the RegisterCodeFixesAsync codepath.  If you can't fix it up (i.e. cs1026 is null), you just bail out.  If you discover you can fix the document, then all you do is capture your modified 'textWithOpenParenthesis' instance, which then makes your callback nothing but: ```c => Task.FromResult(document.WithText(textWithOpenParenthesis))```.

Keeping this text object around should not be very costly as it should share almost all of its data with the real SourceText for the current document.

CyrusNajmabadi commented:
Done with review pass.  Sorry for the added work.

MaStr11 commented:
@CyrusNajmabadi Thanks for the review. The points raised by you made me re-think the approach and I realized that relying on a compiler error to find the position of the closing parenthesis is unnecessarily complicated. Could you please have a look at the changes in commit 70d2add?
I will address all the other issues of your review later but before I would like know whether this still holds true after the changes I made:

> Specifically, i don't like that we can end up trying to fix things, but then might run into a problem half way through the fix that then means we do a partial fix.

The current fix actually inserts at least an opening parenthesis. Therefore there is no need to eagerly calculate the fix in RegisterCodeFixesAsync.

Please give me a short notice if this is fine to you and afterwards I will make all the other corrections.

CyrusNajmabadi commented:
I see.  So you only insert the close paren if you have a parenthesized expression with a missing close paren.  If you don't have a missing close paren then there's no problem returning the doc as is.  Is that right?

If so, i suppose that's a reasonable approach.  :)

CyrusNajmabadi commented:
My only concern with this approach is what happens if there's an extra random ) somewhere in the interpolation expression.  But it's unlikely that that will happen, and i doubt the experience would be a problem.

MaStr11 commented:
Yes. That is the question. I could return the original document or the one with the open parenthesis inserted. If we do the original document there would be a code path that actually does nothing and eager calculation would be an option.

MaStr11 commented:
> My only concern with this approach is what happens if there's an extra random ) somewhere in the interpolation expression.

I have some tests in place to cover this but I couldn't trick the syntax parser.

MaStr11 commented:
Me too. That's why I was asking 😁

MaStr11 commented:
The changes requested by @CyrusNajmabadi are addressed now. Another review would be appreciated.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,80 @@
+            if (conditionalExpression != null)
+            {
+                var documentChangeAction = new MyCodeAction(
+                    cancellationToken => GetChangedDocumentAsync(context.Document, conditionalExpression.SpanStart, cancellationToken));
```
nit: we usually use 'c' for a lambda cancellatoin token.  But don't need to change this if you don't want to.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,80 @@
+            {
+                var parenthesizedExpressionWithClosingParen =
+                    parenthesizedExpression.WithCloseParenToken(SyntaxFactory.Token(SyntaxKind.CloseParenToken)).
+                    WithTriviaFrom(parenthesizedExpression);
```
this is slightly odd trivia-wise.  likely better  to have:  parenthesizedExpression.WithoutTrivia().WithClose(...).WithTriviaFrom(parenthesized).

This prevents a doubling up of trivia which is technically possible here.

MaStr11 replied:
I see. That's why I had problems with some disappearing whitespace. I fixed it slightly different (see [here](https://github.com/dotnet/roslyn/pull/23587/commits/d56ce9eb343264fbaa6bbf7e8e2ab44f878d479f#diff-9a6996bd4fc7eaf5b2a0be89c39f1255R59)).

CyrusNajmabadi commented on src/Features/CSharp/Portable/CodeFixes/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider.cs:
```diff
@@ -0,0 +1,80 @@
+                : base(
+                      CSharpFeaturesResources.Add_parentheses_around_conditional_expression_in_interpolated_string,
+                      createChangedDocument,
+                      CSharpFeaturesResources.Add_parentheses_around_conditional_expression_in_interpolated_string)
```
nit, our convention is to keep the first argument with base(CSharp..., and then align the other args with that first one if they wrap.

But don't need to fix if you don't want to.

sharwell commented on src/Features/CSharp/Portable/CSharpFeaturesResources.Designer.cs:
```diff
@@ -1,10 +1,10 @@
 // <auto-generated>
-//     This code was generated by a tool.
-//     Runtime Version:4.0.30319.42000
+//     Dieser Code wurde von einem Tool generiert.
```
❗️ @tmeschter I have no idea what happened here, but I'm glad I expanded this designer file.

Neme12 replied:
why does this even get localized? that doesn't make any sense...

jmarolf commented:
@sharwell I believe the resource file changes are a consequence of running resgen on a machine set to a german locale (which I believe @MaStr11 is).  We'll just want to revert these changes except for the new string that was added,

tmeschter commented:
I'll see if I can re-generate the .Designer.cs file and update the PR.

MaStr11 commented:
I have a German edition of VS installed and I think [there is nothing I can do to prevent this behavior](https://stackoverflow.com/questions/45015064/force-resgen-exe-to-generate-english-comments-in-the-strongly-typed-resource-cla).

tmeschter commented:
@MaStr11 No worries, I've updated the PR. Note that if you need to make further updates you'll need to pull my commit from your GitHub repo first.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Diagnostics/ConditionalExpressionInStringInterpolation/CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProviderTests.cs:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsAddParenthesesAroundConditionalExpressionInInterpolatedString)]
+        public async Task TestAddParenthesisSimpleConditionalExpression()
```
Nit: pluralize these: TestAddParenthesis

sharwell commented:
@Pilchie This is already based on **dev15.7.x** so it would not be difficult to merge there. Do you have a preference?

Pilchie commented:
This is currently targeting `master`, not `dev15.7.x`.  With that said, I'm open to taking small, low risk things like this for 15.7.

MaStr11 commented:
CI currently fails with `error RS1016: 'CSharpAddParenthesesAroundConditionalExpressionInInterpolatedStringCodeFixProvider' registers one or more code fixes, but does not override the method 'CodeFixProvider.GetFixAllProvider'. Override this method and provide a non-null FixAllProvider for FixAll support, potentially 'WellKnownFixAllProviders.BatchFixer', or 'null' to explicitly disable FixAll support`. 

What is the recommended approach here? Is it to suppress the warning via pragma ([as many analyzers do](https://github.com/dotnet/roslyn/search?q=RS1016)) or should we explicit opt-out by returning `null`. @CyrusNajmabadi and I [discussed](https://github.com/dotnet/roslyn/pull/23587#discussion_r155864902) FixAll and came to the conclusion that there is no need to offer it:

> @CyrusNajmabadi 
> Derive from SyntaxEditorBasedCodeFixProvider. That way you can easily supply a fast and efficient 'FixAll' experience.
> 
> @Mastr11
> FixAll is IMHO not useful for this kind of error (this fixes a syntax error and I don't see how there could be more then one instance of this error at a time). If there are other reasons I will change the base class (or if you just insists).
> 
> @CyrusNajmabadi 
> You're def right that fix-all is less useful here. where it could be valuable is for someone who changes a bunch of code, or copy pastes a lot. But if you don't want a fix-all, that's fine :)

sharwell commented:
> What is the recommended approach here? Is it to suppress the warning via pragma (as many analyzers do) or should we explicit opt-out by returning null.

Explicitly opt-out by returning null. You can include a summary of the decision in a comment.

sharwell commented:
@Pilchie 😄 (and feel free to merge if approved)
